### DISCUSSION TITLE
libvips JVM bindings using new FFM API

### DISCUSSION THREAD

**Discussion by lopcode** [libvips/libvips#4112] [General]
Hi there,

Thanks for working on libvips!

This is just a heads up that I shipped Java/JVM libvips bindings that use Java's new FFM API (Foreign Function & Memory API) from [JEP 454](https://openjdk.org/jeps/454). The bindings are automatically generated from headers (via a new JDK tool called [jextract](https://github.com/openjdk/jextract/)), and then again automatically processed with some custom code. The second pass makes a "helper API" that is safer to use than the "raw" C stuff (as noted on your website), and more idiomatic for JVM systems. In theory there should be ~100% API coverage, assuming Java's generation tooling picked everything up, and updating for new libvips releases should be easy.

No doubt there are additional improvements to make, but I think things are good enough to use now. The project's over here if you'd like to take a look: https://github.com/lopcode/vips-ffm

I was wondering if you would consider signposting it on the libvips website? No problem if not, and I know there are other more mature JNI-based solutions available too. But I thought the new project might be useful for others, so wanted to reach out and ask.

Thanks again,
lopcode

**Comment by jcupitt** [comment]
Hi @lopcode, sure, I'll add a note, it looks nice. 

It seems to be generating off the C++ headers, is that right?

I suppose (in my mind) the ideal JVM binding would target the lower-level libvips GObject API, rather as [the C# binding](https://github.com/kleisauke/net-vips) does. It would make the whole libvips API available, wouldn't need the (often rather fragile) header parsing, and would automatically update itself for whatever libvips binary it detected at runtime, rather than being tied to a specific libvips version. I'm sure you saw https://www.libvips.org/API/current/binding.html

However! That's quite a bit more work, and I can see that pragmatically your approach is easier. 

Thanks for letting us know!

**Reply by jcupitt** [reply]
Here's the bit of netvips that calls a libvips operation:

https://github.com/kleisauke/net-vips/blob/master/src/NetVips/Operation.cs#L138-L254

That uses libvips introspection to call any operation in the libvips binary, passing and converting arguments back and forth. There's similar code in the C++, python, ruby, php, lua, etc. bindings.

**Comment by lopcode** [comment]
Thanks for looking, appreciate you adding a note, and for the pointer to the C# bindings.

The main benefit of targeting the GObject API seems to be "automatic" updating for API changes which does sound appealing. I did see the bindings section on the website and went for something closer to the "Dynamic language with FFI" option, except `jextract` means there's no runtime "reflection" penalty. Though - even with a dynamic approach you're probably only paying that cost on the first method call in a JVM system.

I'll play around with it, thanks again 👍

**Reply by jcupitt** [reply]
The nice thing about the gobject API is that it decouples the binding from the libvips version. I think your JVM binding would potentially segv if the libvips binary it linked to at runtime didn't match the libvips headers that were used to generate the binding -- your binding is tied to a libvips version.

You're right, there's a runtime cost to that, though of course your JVM binding is to the C++ API, which internally uses libvips reflection to call into the gobject API, so you're probably (invisibly!) paying the penalty anyway.

cffi (the python FFI module) has an interesting way around this -- they support libraries in ABI and API mode. ABI mode just uses `ldopen()` plus some python to construct C stack frames for calls in the way you'd expect. API mode is the clever one -- at install time, it tries to generate, compile and install a C module which has wrappers for all calls into the library, in effect replacing the use of FFI. Moving argument marshaling from python to C gives a nice speedup, and makes integration with the python GC a lot simpler (esp. for callbacks).

**Reply by jcupitt** [reply]
> I'll play around with it, thanks again 👍

No problem, let me know how you get on!

Shall I add a note to the libvips README now, or would you like some more time?

**Reply by lopcode** [reply]
Now sounds good to me - I think the API is good enough to use as-is, and I don't mind supporting a future change if there is one

**Reply by jcupitt** [reply]
OK, I added a link. Do let me know if you do more on your project!

**Reply by lopcode** [reply]
I've now implemented your suggestion, and vips-ffm uses the GObject API to generate bindings 🎉

For interest, this is the guts of the "call" equivalent to perform an operation: https://github.com/lopcode/vips-ffm/blob/main/core/src/main/java/app/photofox/vipsffm/VipsInvoker.java#L18 - a lot of the complexity is just converting between Java types and GObject/Vips types. Usage ends up being a lot nicer with a `VImage` "box type" like the C++ bindings: https://github.com/lopcode/vips-ffm/blob/main/sample/src/main/kotlin/vipsffm/VImageCreateThumbnailSample.kt#L14

A bonus is that it's easy to generate inline documentation from the information in the GObject API, including for optional arguments: https://github.com/lopcode/vips-ffm/blob/main/core/src/main/java/app/photofox/vipsffm/VImage.java#L5788

Still some rough edges to figure out, like some jextract limitations causing enums to be represented as raw integers - but that's a problem for another time. 

Thanks for pointing me in the right direction 🙂

**Reply by jcupitt** [reply]
That's so cool! I read your Operation caller, it's rather neat.

> limitations causing enums to be represented as raw integers

Most of the bindings accept strings as well for enum values, then look them up as the "nickname" of the GEnum. You can write eg.:

```C++
    VImage im = VImage::new_from_file(argv[i],
        VImage::option()->set("access", "sequential"));
```

Errors won't be caught at compile time, of course, but it is very convenient.

**Reply by jcupitt** [reply]
I like your docs too. There are more complete docs in `Vips-8.0.gir`, did you see? Eg.:

```xml
    <function name="thumbnail"
              c:identifier="vips_thumbnail"
              introspectable="0">
      <doc xml:space="preserve"
           filename="libvips/resample/thumbnail.c"
           line="1210">Optional arguments:

* @height: %gint, target height in pixels
* @size: #VipsSize, upsize, downsize, both or force
* @no_rotate: %gboolean, don't rotate upright using orientation tag
* @crop: #VipsInteresting, shrink and crop to fill target
* @linear: %gboolean, perform shrink in linear light
* @import_profile: %gchararray, fallback import ICC profile
* @export_profile: %gchararray, export ICC profile
* @intent: #VipsIntent, rendering intent
* @fail_on: #VipsFailOn, load error types to fail on

Make a thumbnail from a file. Shrinking is done in three stages: using any
shrink-on-load features available in the file import library, using a block
shrink, and using a lanczos3 shrink. At least the final 200% is done with
lanczos3. The output should be high quality, and the operation should be
quick.

See vips_thumbnail_buffer() to thumbnail from a memory buffer, or
vips_thumbnail_source() to thumbnail from an arbitrary byte source.

By default, libvips will only use the first frame of animated or multipage
images. To thumbnail all pages or frames, pass `n=-1` to the loader in
@filename, for example `"x.gif[n=-1]"`.
... snip
See also: vips_thumbnail_buffer().</doc>
      <source-position filename="libvips/include/vips/resample.h" line="82"/>
      <return-value transfer-ownership="none">
        <doc xml:space="preserve"
             filename="libvips/resample/thumbnail.c"
             line="1283">0 on success, -1 on error.</doc>
        <type name="gint" c:type="int"/>
      </return-value>
      <parameters>
        <parameter name="filename" transfer-ownership="none">
          <doc xml:space="preserve"
               filename="libvips/resample/thumbnail.c"
               line="1212">file to read from</doc>
          <type name="utf8" c:type="const char*"/>
        </parameter>
        <parameter name="out"
                   direction="out"
                   caller-allocates="0"
                   transfer-ownership="full">
          <doc xml:space="preserve"
               filename="libvips/resample/thumbnail.c"
               line="1213">output image</doc>
          <type name="Image" c:type="VipsImage**"/>
        </parameter>
        <parameter name="width" transfer-ownership="none">
          <doc xml:space="preserve"
               filename="libvips/resample/thumbnail.c"
               line="1214">target width in pixels</doc>
          <type name="gint" c:type="int"/>
        </parameter>
        <parameter name="..." transfer-ownership="none">
          <doc xml:space="preserve"
               filename="libvips/resample/thumbnail.c"
               line="1215">%NULL-terminated list of optional named arguments</doc>
          <varargs/>
        </parameter>
      </parameters>
    </function>
```

It wouldn't be hard to add some of that to your doc generator too.

**Reply by lopcode** [reply]
I shipped GIR-based docs today - it ended up being a bit fiddly to massage them in to a format compatible with Javadoc. But references are nicely integrated, and having these available in your IDE is very helful. For example: https://github.com/lopcode/vips-ffm/blob/main/core/src/main/java/app/photofox/vipsffm/VImage.java#L2665

**Reply by jcupitt** [reply]
Wow, that looks nice! Do you have a copy of the formatted docs somewhere? I'd be interested to see.

**Reply by lopcode** [reply]
They're turned directly from GTKDoc to Javadoc with some regular expressions - most of it is fine, but there were some fixups needed at the end (especially relating to tables tables) needed to make them valid: https://github.com/lopcode/vips-ffm/blob/main/generator/src/main/kotlin/vipsffm/GenerateVClasses.kt#L290

**Reply by jcupitt** [reply]
Sorry, I mean formatted HTML docs I could read somewhere. I was just wondering what they look like, how links work, that kind of thing.

Other bindings have a copy of the formatted docs in github pages, for example:

http://libvips.github.io/ruby-vips/

**Reply by lopcode** [reply]
Ah I misunderstood! There's no website or rendered HTML docs at the moment - something I need to get around to 🙂

**Reply by lopcode** [reply]
This was easier than I was expecting with Javadoc, especially now it supports Markdown so I can use the GitHub README as the "website overview".

The VImage docs for example: https://vipsffm.photofox.app/app.photofox.vipsffm/app/photofox/vipsffm/VImage.html

**Reply by jcupitt** [reply]
The GIR docs drop in well! It's very close!

It's a shame gtk-doc markdown doesn't support tables directly. The in-line HTML tables end up as text :( 

https://vipsffm.photofox.app/app.photofox.vipsffm/app/photofox/vipsffm/VImage.html#add(app.photofox.vipsffm.VipsOption...)

I'm sure you don't need more suggestions, but just in case ... the C++ binding has a set of extra hand-written convenience methods, perhaps you saw:

https://github.com/libvips/libvips/blob/master/cplusplus/include/vips/VImage8.h#L1119

Adding something like that would you write code like:

```java
c = a.less(12).ifthenelse(14, b);
```

Which would find pixels in `a` less than 12 and set them to 14, taking other pixels from `b`.

**Comment by lopcode** [comment]
I've promoted the most recent build to `v1.0.0` and I'll close this discussion now, thanks again 🙂 

Also discovered that JVips is quite out of date at the moment (2 years at least) - I had planned to benchmark against a recent release, but it seems there's binary compatibility issues between it and recent libvips releases.

**Reply by jcupitt** [reply]
Great! You must be pleased.

I hadn't realised jvips was languishing :( that's a shame.
